p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,350 @@
+	localFingerprintMultibase string
+
+	localAddr      net.Addr
+	localMultiaddr ma.Multiaddr
+
+	// buffered incoming connections
+	acceptQueue chan tpt.CapableConn
+
+	// Accepting a connection requires instantiating a peerconnection
+	// and a noise connection which is expensive. We therefore limit
+	// the number of in-flight connection requests. A connection
+	// is considered to be in flight from the instant it is handled
+	// until it is dequed by a call to Accept, or errors out in some
+	// way.
+	inFlightInputQueue     chan candidateAddr
+	maxInFlightConnections uint32
+
+	// used to control the lifecycle of the listener
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func newListener(transport *WebRTCTransport, laddr ma.Multiaddr, socket net.PacketConn, config webrtc.Configuration) (*listener, error) {
+	localFingerprints, err := config.Certificates[0].GetFingerprints()
+	if err != nil {
+		return nil, err
+	}
+
+	localMh, err := hex.DecodeString(strings.ReplaceAll(localFingerprints[0].Value, ":", ""))
+	if err != nil {
+		return nil, err
+	}
+	localMhBuf, err := multihash.Encode(localMh, multihash.SHA2_256)
+	if err != nil {
+		return nil, err
+	}
+	localFpMultibase, err := multibase.Encode(multibase.Base64url, localMhBuf)
+	if err != nil {
+		return nil, err
+	}
+
+	inFlightQueueCh := make(chan candidateAddr, 1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	mux := udpmux.NewUDPMux(socket, func(ufrag string, addr net.Addr) bool {
+		// Push to the inFlightQueue asynchronously to avoid blocking the mux goroutine
+		// on candidates being processed. This can cause new connections to fail at high
+		// throughput but will allow packets for existing connections to be processed.
+		select {
+		case inFlightQueueCh <- candidateAddr{ufrag: ufrag, raddr: addr.(*net.UDPAddr)}:
+			return true
+		default:
+			log.Debug("candidate chan full, dropping incoming candidate")
+			return false
+		}
+	})
+
+	l := &listener{
+		mux:                       mux,
+		transport:                 transport,
+		config:                    config,
+		localFingerprint:          localFingerprints[0],
+		localFingerprintMultibase: localFpMultibase,
+		localMultiaddr:            laddr,
+		ctx:                       ctx,
+		cancel:                    cancel,
+		localAddr:                 socket.LocalAddr(),
+		acceptQueue:               make(chan tpt.CapableConn, transport.maxInFlightConnections-1),
+		inFlightInputQueue:        inFlightQueueCh,
+		maxInFlightConnections:    transport.maxInFlightConnections,
+	}
+
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+		l.inFlightWorker()
+	}()
+
+	return l, err
+}
+
+func (l *listener) inFlightWorker() {
+	for {
+		select {
+		case <-l.ctx.Done():
+			return
+
+		case addr := <-l.inFlightInputQueue:
+			ctx, cancel := context.WithTimeout(l.ctx, candidateSetupTimeout)
+			defer cancel()
+
+			conn, err := l.handleCandidate(ctx, &addr)
```

Comment:
Do I understand correctly that this blocks until the connection has been fully set up? That would be around 1 RTT (the duration of the Noise handshake), which means that we can only accept a single connection per RTT.

Reply:
You're right. In my initial design this was not the case, this is the result of the last refactor.
In the initial design there were 10 workers, each which takes care of their own connection to be setup.
You said however you did not like that, so reduced it instead to a single worker, while still limiting the pipeline to a max of 10 in the overall queue. Open to other ideas if not any of these 2.